cli: Add validator check for Dataset name

scallaway opened the pull request:
Adds validation to input for the Dataset name to ensure that the URL that's constructed is what the API expects.

davidhewitt commented on cli/src/commands/create/dataset.rs:
```diff
@@ -1,15 +1,30 @@
+
+/// Ensures the name provided conforms to the <owner>/<name> structure
+fn validate_dataset_name(s: String) -> std::result::Result<(), String> {
+    match RX_FULL_NAME.is_match(&s) {
+        true => Ok(()),
+        false => Err(s),
+    }
```
I personally prefer `if x { foo } else { bar }` rather than matching on boolean.

Also possible is to use something along the lines of `failchain::ensure!(RX_FULL_NAME.is_match(&s), ...)` as you're returning an error on the bad case - but that's probably not very ergonomic until we migrate from `failchain` to `anyhow`.

scallaway replied:
Yep, makes sense to just use an `if` here rather than a `match`.

I don't think that we can use `anyhow` here since the expected return type of a validator function for `structopt` is `Result<(), String>` and so a conversion to an `anyhow` Error doesn't work here.

davidhewitt commented on cli/src/commands/create/dataset.rs:
```diff
@@ -1,15 +1,30 @@
 
 use crate::errors::{ErrorKind, Result};
 
+lazy_static! {
+    static ref RX_FULL_NAME: Regex =
+        Regex::new("^[A-Za-z0-9-_.]{1,1024}/[A-Za-z0-9-_]{1,256}").unwrap();
+}
```
Optional: we're tending towards using `once_cell` instead of `lazy_static`: no need for macro magic and also `once_cell` is potentially on the way to the standard library.

Could use `once_cell::sync::Lazy` if you fancy trying it out. (Or indeed migrate the whole cli project to `once_cell` in a separate PR?)

scallaway replied:
I think converting this here to use `once_cell` rather than `lazy_static` is a good idea, and I'll do it in this PR.

Should probably leave conversion of the rest of the project to `once_cell` to a different PR like you say!

davidhewitt reviewed (approved):
Thanks, a nice usability improvement! Optional if you think this is worth mentioning in the changelog.

Also two super nitty suggestions below.

davidhewitt commented:
Also, if you're up for a little extra challenge: add a test in `cli/tests/` for this. There's a number of tests in that directory which execute the cli binary and verify its output. Here you could run with invalid name and verify it gives the helpful error message. (Could even add a test to successfully create a dataset, if you're up for it!)

tommilligan commented on cli/src/commands/create/dataset.rs:
```diff
@@ -1,15 +1,30 @@
 
+lazy_static! {
+    static ref RX_FULL_NAME: Regex =
+        Regex::new("^[A-Za-z0-9-_.]{1,1024}/[A-Za-z0-9-_]{1,256}").unwrap();
```
I thought we would want to limit this to a shorter name overall due to API constraints? cc: @annesylvie and her recent api validation work.

I'm also happy to leave this more permissible and just return the API error if the name is too long.

tommilligan replied:
Currently, with this regex you can get an error like:
```bash
$ cargo run -- create dataset org1/some-incredibly-long-dataset-name-that-will-cause-a-500-error-in-the-api-because-its-over-127-chars-long-no-really-now-its-over-the-limit
E Operation to create a dataset has failed.
E  |- API request failed with 422 Unprocessable Entity: Dataset name is too long; the maximum length of the string org1/<dataset_name> is of 127 characters
```
which I think is absolutely fine

scallaway replied:
Maybe I combined the wrong regex? I pulled this from the API code, happy to update it to something even more restrictive if that's the way to go.

annesylvie replied:
I did indeed add a new constraint on the length of the string `<org>/<dataset_name>` that is accepted by the api; I guess this was merged after you started working on this!

The regex itself is unchanged; we check for the length of the string separately. Up to you if you also want to implement this behaviour @scallaway. It can also be another PR if you prefer.

tommilligan replied:
I think in that case it's fine to leave the regex as is. We expect to surface API errors directly to users in a lot of cases, the validation in the CLI should be expected to be light (similar to frontend validation)

scallaway commented:
@tommilligan @davidhewitt 

Re-requesting reviews from both of you as I've added a new file `test_datasets` since that didn't exist previously, and modeled it based off `test_sources`. Plus I've addressed the suggestions from David!

tommilligan reviewed (approved):
LGTM, thanks for doing the tests

davidhewitt reviewed (approved):
Looks great. Thanks very much for improving this!
